### ISSUE TITLE
Mapbox Immutable Style with Immer

### ISSUE THREAD

**Issue by krewllobster** [uber/react-map-gl#924]
I am using Redux-starter-kit, which provides immutability by default in its reducers utilizing the Immer library (as opposed to Immutable.js). However, when running in development mode, Immer freezes all objects created with its producer function. If you try to create an immutable style for the map using immer, when it receives the updated version it throws the following error: 

```
style-utils.js:18 Uncaught TypeError: Cannot assign to read only property 'layers' of object '#<Object>'
    at normalizeStyle (style-utils.js:18)
    at StaticMap.componentDidMount (static-map.js:160)
    at StaticMap.componentDidMount (react-hot-loader.development.js:714)
    at commitLifeCycles (react-dom.development.js:21933)
    at commitLayoutEffects (react-dom.development.js:25266)
    at HTMLUnknownElement.callCallback (react-dom.development.js:363)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:412)
    at invokeGuardedCallback (react-dom.development.js:465)
    at commitRootImpl (react-dom.development.js:25002)
    at unstable_runWithPriority (scheduler.development.js:818)
    at runWithPriority$2 (react-dom.development.js:12259)
    at commitRoot (react-dom.development.js:24843)
    at finishSyncRender (react-dom.development.js:24220)
    at performSyncWorkOnRoot (react-dom.development.js:24194)
    at react-dom.development.js:12313
    at unstable_runWithPriority (scheduler.development.js:818)
    at runWithPriority$2 (react-dom.development.js:12259)
    at flushSyncCallbackQueueImpl (react-dom.development.js:12308)
    at flushSyncCallbackQueue (react-dom.development.js:12296)
    at unbatchedUpdates (react-dom.development.js:24351)
    at legacyRenderSubtreeIntoContainer (react-dom.development.js:27638)
    at render (react-dom.development.js:27729)
    at Module../src/index.js (index.js:51)
    at __webpack_require__ (bootstrap:785)
    at fn (bootstrap:150)
    at Object.0 (mapSelectors.js:5)
    at __webpack_require__ (bootstrap:785)
    at checkDeferredModules (bootstrap:45)
    at Array.webpackJsonpCallback [as push] (bootstrap:32)
    at main.chunk.js:1
```

It seems as if the normalize style/style diffing parts of the library are attempting to mutate the mapStyle that is granted from the component, which is obviously not possible to do with a frozen object. 

This error persists even when "preventStyleDiffing" is set to true in the component.

Should the map be mutating the style object? Am I missing something super simple? Or do I just need to fall back to using Immutable.js for the map style, and Immer for everything else?

I am using:
```
react-map-gl: "^5.0.12"
redux-starter-kit: "^1.0.0"
```
And redux-starter-kit relies on:
```
immer: "^4.0.1"
redux: "^4.0.0"
```

**Comment by krewllobster** [comment]
Related to the above, Immer does not allow storing the ViewState because the viewState given to the onViewStateChange includes non-serializable values like transitionEasing (function) and transitionInterpolator (LinearInterpolator class instance).

React's build-in useState has no problems with it though.

**Comment by markerikson** [comment]
Correct.  It's possible to customize the serializability check middleware to whitelist those sections, but as a general principle this is bad behavior.

**Comment by krewllobster** [comment]
(@markerikson authors & maintains redux-starter-kit)

**Comment by Pessimistress** [comment]
This should be fixed in ^5.1.0.
